support/ref-docs/src/documenter.ts

```diff
@@ -0,0 +1,624 @@
+import {
+  root,
+  text,
+  list,
+  listItem,
+  paragraph,
+  strong,
+  brk,
+  link,
+  inlineCode,
+  table,
+  tableRow,
+  tableCell,
+  code,
+  html
+} from 'mdast-builder';
+import {
+  ApiClass,
+  ApiConstructor,
+  ApiConstructSignature,
+  ApiDocumentedItem,
+  ApiEnum,
+  ApiFunction,
+  ApiInterface,
+  ApiItemKind,
+  ApiMethod,
+  ApiMethodSignature,
+  ApiModel,
+  ApiPackage,
+  ApiParameterListMixin,
+  ApiPropertyItem,
+  ApiTypeAlias,
+  Parameter,
+  ApiDeclaredItem,
+  Excerpt,
+  ApiEntryPoint,
+  ExcerptTokenKind
+} from '@microsoft/api-extractor-model';
+import {
+  DocSection,
+  DocParagraph,
+  DocPlainText,
+  DocLinkTag,
+  DocSoftBreak,
+  DocCodeSpan,
+  DocBlock
+} from '@microsoft/tsdoc';
+import { Node, Parent } from 'unist';
+import { flatMap, joinArrs, filterUndef } from './arrays';
+import {
+  section,
+  itemSections,
+  Ctor,
+  itemsOfType,
+  SectionTitle
+} from './macros';
+
+/**
+ * Information about the context of a documentation process.
+ */
+export interface DocumentationContext {
+  model: ApiModel;
+  pkg: ApiPackage;
+  pkgEntry: ApiEntryPoint;
+}
+
+/**
+ * Binds a function to a documentation context.
+ */
+function withCtx<
+  F extends (ctx: DocumentationContext, ...args: A) => R,
+  A extends any[],
+  R
+>(ctx: DocumentationContext, fn: F): (...args: A) => R {
+  return (...args: A) => fn(ctx, ...args);
+}
+
+/**
+ * Heuristic that decides if a parameter utilises object destructuring.
+ */
+function isDestructuredParam(param: Parameter): boolean {
+  return param.name.startsWith('{') && param.name.endsWith('}');
+}
+
+/**
+ * Heuristic that decides if a parameter is required or not
+ */
+function isRequiredParam(item: ApiDeclaredItem, param: Parameter): boolean {
+  // If a `?` is found in the excerpt token before the type, we assume that this
+  // parameter is optional, e.g. `?: ` before the token `string`
+  return !item.excerptTokens[
+    param.parameterTypeExcerpt.tokenRange.startIndex - 1
+  ].text.includes('?');
+}
+
+/**
+ * Prepends a list of child nodes to the first found paragraph. If no paragraph
+ * is found, the children aren't inserted.
+ */
+function prependFirstParagraph(
+  nodes: readonly Node[],
+  children: readonly Node[]
+): Node[] {
+  // Find the first paragraph
+  const firstPara = nodes.find(node => node.type === 'paragraph');
+  if (firstPara !== undefined) {
+    // Prepend the new children
+    (firstPara as Parent).children.unshift(...children);
+  }
+
+  return nodes as Node[];
+}
+
+/**
+ * Attempts to link to a target API item by its name.
+ */
+function tryLinkName(
+  ctx: DocumentationContext,
+  name: string,
+  child: Node
+): Node {
+  // Attempt to find the target item
+  const [target] = ctx.pkgEntry.findMembersByName(name);
+  // TODO: handle multiple targets?
+  // TODO: better extraction by using tokens instead of
+
+  // If the target isn't found, simply return the child w/o a link
+  if (target === undefined) {
+    return child;
```

Comment:
does this mean that when the target isn't found, the node will link to itself? i'm wondering if returning `undefined` would keep things simpler.

Reply:
If it can't find a target API item, it returns the node (text, inline code) that is displayed inside of the `<a>` tag/`[]` markdown brackets.

That means if you get `tryLinkName(..., 'thisNameDoesntExist', text('thisNameDoesntExit'))` (the last element being an AST `text` node) then it'd just return the `text` node instead of a `link` node with the `text` node as its child.

So it doesn't link to anything because it won't be a link.